### PR TITLE
T12251 -- Add suport for Dropbox

### PR DESCRIPTION
**Pull request by xaiki**
in this PR we first reorganize the external-apps plugin a bit so that we can easily add more fields to the json, then implement a tar extractor (with a tar json type).

that's all folks.

https://phabricator.endlessm.com/T12251

### CODE REVIEW

**Comment by xaiki** [comment]
commits squashed and properly commented.

File: src/plugins/gs-plugin-external-apps.c [hunk 1]
```diff
@@ -296,6 +296,23 @@ add_runtime_deb_asset (const gchar *build_dir,
...
 static gboolean
+add_runtime_tar_asset (const gchar *build_dir,
+		       const gchar *repo_dir,
+		       const char *asset_path,
+		       GError **error)
+{
+	/* flatpak build --runtime needs files in /usr, when coming from a
+	   debian package we can asume some files in /usr, when coming from
+	   a tgz, it's harder, so force that here. */
+	g_autofree gchar *extract_path = g_strdup_printf ("%s/usr", repo_dir);
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Use g_build_filename.

File: src/plugins/gs-plugin-external-apps.c [hunk 2]
```diff
@@ -43,8 +44,15 @@
 
 #define TMP_ASSSETS_PREFIX "gs-external-apps"
 
+enum GsPluginExternalType {
+	GS_PLUGIN_EXTERNAL_TYPE_DEB = 1,
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Why starting with 1?

File: src/plugins/gs-plugin-external-apps.c [hunk 3]
```diff
@@ -43,8 +44,15 @@
...
+	GS_PLUGIN_EXTERNAL_TYPE_TAR,
+	GS_PLUGIN_EXTERNAL_TYPE_UNKNOWN,
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Extra comma here?
Also, I wonder whether the TYPE_UNKNOWN should come as the first type, so it's value doesn't keep changing (assuming you'd keep UNKNOWN as the last type).

File: src/plugins/gs-plugin-external-apps.c [hunk 4]
```diff
@@ -43,8 +44,15 @@
...
+};
+
 struct GsPluginData {
 	GsFlatpak	*flatpak;
+	enum GsPluginExternalType type;
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: I'd prefer that you typedef'ed this type in the previous enum declaration. So we'd skip using *enum* here.

File: src/plugins/gs-plugin-external-apps.c [hunk 5]
```diff
@@ -514,6 +525,17 @@ extract_runtime_info_from_json_data (GsPlugin *plugin,
 		return FALSE;
 	}
 
+	/* optional elements */
+	type = json_object_get_string_member (runtime, JSON_RUNTIME_TYPE_KEY);
+	if (type) {
+		if (g_strcmp0(type, "deb") == 0)
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Space needed between function name and parenthesis. Same below.

**Comment by joaquimrocha** [comment]
@xaiki  I don't understand why the first commit is needed here (96e2153).
I understand that if we need to extract a bunch of data fields from the json we should use a structure instead of output params, but I don't like the idea of adding data that is related to one individual runtime to a plugin struct that is general.

If you want to keep things more organized, as I mentioned above, please create an asset/runtime struct to represent that object.

File: src/plugins/gs-plugin-external-apps.c (hunk 2 above)
Comment (xaiki) [inline src/plugins/gs-plugin-external-apps.c]: so that 0 initialized is caught.

File: src/plugins/gs-plugin-external-apps.c (hunk 3 above)
Comment (xaiki) [inline src/plugins/gs-plugin-external-apps.c]: that allows me to test for >= UNKNOWN

File: src/plugins/gs-plugin-external-apps.c (hunk 4 above)
Comment (xaiki) [inline src/plugins/gs-plugin-external-apps.c]: done.

File: src/plugins/gs-plugin-external-apps.c (hunk 1 above)
Comment (xaiki) [inline src/plugins/gs-plugin-external-apps.c]: done

**Comment by xaiki** [comment]
why add complexity with yet another struct that we need to unpack from priv, as our priv is pretty small ?

File: src/plugins/gs-plugin-external-apps.c (hunk 2 above)
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Then that's one more reason why UNKNOWN should come first IMO.

File: src/plugins/gs-plugin-external-apps.c (hunk 3 above)
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Why would you want to test it like that? I don't think enums should be tested with < or > .
Mind the comma too.

**Comment by joaquimrocha** [comment]
Because of the reasons I've stated before + getting the code less messy. I'd likely not even add this extra structure to priv but just pass it around if feasible.

**Comment by xaiki** [comment]
i'll ask @cosimoc to cut on this, we clearly have oposite views of what's messy.

**Comment by joaquimrocha** [comment]
Sure, let's have @cosimoc untie this.
FTR, here's why I think it's messy to have those attributes in the plugin's priv:
1) They belong to the runtime extension, not the plugin;
2) You're using a permanent struct (the plugin's priv) to store values that will be temporary and depending on each runtime used (this is error prone);
3) You're just adding them to the priv because it is a bit laborious to keep passing multiple arguments so you accomplish the same with my proposal;
4) If you need 5 more attributes you'll need to add them all to the priv, making it messier in this case alone.

**Comment by cosimoc** [comment]
I have to agree with @joaquimrocha here, since I had the same comment when reviewing the code yesterday...

Basically that data is only valid in the context of the code that builds the external runtime, not in the context of the whole plugin, which is the owner of the private struct. If another caller was to call into the plugin and see/use the data that was stored in there from a previous installation, that would be wrong and could lead to trouble.

File: src/plugins/gs-plugin-external-apps.c [hunk 6]
```diff
@@ -283,19 +291,54 @@ add_runtime_deb_asset (const gchar *build_dir,
 	return TRUE;
 }
 
+GsPluginExternalType
+get_type_from_string (const gchar *type)
+{
+	if (type) {
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: You use `g_strcmp0()` below, so there's no need to check for `NULL` here

File: src/plugins/gs-plugin-external-apps.c [hunk 7]
```diff
@@ -283,19 +291,54 @@ add_runtime_deb_asset (const gchar *build_dir,
...
+		if (g_strcmp0(type, "deb") == 0)
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: Space before paren

File: src/plugins/gs-plugin-external-apps.c [hunk 8]
```diff
@@ -283,19 +291,54 @@ add_runtime_deb_asset (const gchar *build_dir,
...
+add_runtime_tar_asset (const gchar *build_dir,
+		       const gchar *repo_dir,
+		       const char *asset_path,
+		       GError **error)
+{
+	/* flatpak build --runtime needs files in /usr, when coming from a
+	   debian package we can asume some files in /usr, when coming from
+	   a tgz, it's harder, so force that here. */
+	g_autofree gchar *extract_path =
+		g_build_filename(repo_dir, "usr", NULL);
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: Space before paren

File: src/plugins/gs-plugin-external-apps.c [hunk 9]
```diff
@@ -283,19 +291,54 @@ add_runtime_deb_asset (const gchar *build_dir,
...
+}
+
+static gboolean
+add_runtime_tar_asset (const gchar *build_dir,
+		       const gchar *repo_dir,
+		       const char *asset_path,
+		       GError **error)
+{
+	/* flatpak build --runtime needs files in /usr, when coming from a
+	   debian package we can asume some files in /usr, when coming from
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: `assume`

File: src/plugins/gs-plugin-external-apps.c [hunk 10]
```diff
@@ -283,19 +291,54 @@ add_runtime_deb_asset (const gchar *build_dir,
...
+		   const char *archive_type,
 		   const char *asset,
 		   GError **error)
 {
 	g_autofree char *content_type = NULL;
 	g_autoptr(GFile) download_file = NULL;
 	g_autofree char *download_name = download_asset (plugin, app, runtime,
 							 asset, error);
+	GsPluginExternalType type =
+		get_type_from_string(archive_type);
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: Space before paren -- also can be on the same line probably

File: src/plugins/gs-plugin-external-apps.c [hunk 11]
```diff
@@ -309,11 +352,18 @@ add_runtime_asset (GsPlugin *plugin,
...
 	g_debug ("Adding runtime asset with type '%s'", content_type);
 
-	if (g_content_type_is_a (content_type, "application/x-deb")) {
+	if ((type == GS_PLUGIN_EXTERNAL_TYPE_DEB) ||
+	    g_content_type_is_a (content_type, "application/x-deb")) {
 		return add_runtime_deb_asset (build_dir, repo_dir,
 					      download_name, error);
 	}
 
+	if ((priv->type == GS_PLUGIN_EXTERNAL_TYPE_TAR) ||
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: Should be `type`, not `priv->type`

File: src/plugins/gs-plugin-external-apps.c [hunk 12]
```diff
@@ -368,6 +418,8 @@ build_runtime (GsPlugin *plugin,
 	const char *runtime_name = gs_app_get_id_no_prefix (runtime);
 	const char *runtime_url = gs_app_get_metadata_item (runtime,
 							    METADATA_URL);
+	const char *runtime_type = gs_app_get_metadata_item(runtime,
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: Space before paren

File: src/plugins/gs-plugin-external-apps.c [hunk 13]
```diff
@@ -509,6 +563,12 @@ extract_runtime_info_from_json_data (const char *data,
 		return NULL;
 	}
 
+	/* optional elements */
+	type_str = json_object_get_string_member (runtime, JSON_RUNTIME_TYPE_KEY);
+	if (type_str) {
```
Comment (cosimoc) [inline src/plugins/gs-plugin-external-apps.c]: `g_strdup()` works just fine with `NULL`, so I think you don't need to check for this here.

File: src/plugins/gs-plugin-external-apps.c [hunk 14]
```diff
@@ -294,17 +294,34 @@ add_runtime_deb_asset (const gchar *build_dir,
 static inline GsPluginExternalType
 get_type_from_string (const char *type)
 {
-	if (type) {
-		if (g_strcmp0(type, "deb") == 0)
-			return GS_PLUGIN_EXTERNAL_TYPE_DEB;
-               else if (g_strcmp0(type, "tar") == 0)
-                       return  GS_PLUGIN_EXTERNAL_TYPE_TAR;
-	}
+
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Unneeded newline here.

File: src/plugins/gs-plugin-external-apps.c [hunk 15]
```diff
@@ -294,17 +294,34 @@ add_runtime_deb_asset (const gchar *build_dir,
...
+add_runtime_tar_asset (const gchar *build_dir,
+		       const gchar *repo_dir,
+		       const char *asset_path,
+		       GError **error)
+{
+	/* flatpak build --runtime needs files in /usr, when coming from a
+	   debian package we can assume some files in /usr, when coming from
+	   a tgz, it's harder, so force that here. */
+	g_autofree gchar *extract_path = g_build_filename (repo_dir, "usr", NULL);
+
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: Unneeded newline here.

File: src/plugins/gs-plugin-external-apps.c [hunk 16]
```diff
@@ -393,7 +416,7 @@ build_runtime (GsPlugin *plugin,
 	const char *runtime_name = gs_app_get_id_no_prefix (runtime);
 	const char *runtime_url = gs_app_get_metadata_item (runtime,
 							    METADATA_URL);
-	const char *runtime_type = gs_app_get_metadata_item(runtime,
+	const char *runtime_type = gs_app_get_metadata_item (runtime,
 							    METADATA_TYPE);
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: This fix doesn't belong in this commit but rather in the one where Cosimo commented on.
Also, please verify that METADATA_TYPE is well-aligned after the new space introduction (I am not sure if it's right or wrong here due to the mix of tabs and spaces that g-s uses..).

File: src/plugins/gs-plugin-external-apps.c [hunk 17]
```diff
@@ -540,9 +563,7 @@ extract_runtime_info_from_json_data (const char *data,
 
 	/* optional elements */
 	type_str = json_object_get_string_member (runtime, JSON_RUNTIME_TYPE_KEY);
-	if (type_str) {
-		*type = g_strdup (type_str);
-	}
+	*type = g_strdup (type_str);
```
Comment (joaquimrocha) [inline src/plugins/gs-plugin-external-apps.c]: This fix also doesn't belong to this commit.
If you need help with git sorcery let me know on Slack. I'll be happy to help.
